platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/hash/MerkleHashBuilder.java

```diff
@@ -142,77 +113,104 @@ public Future<Hash> digestTreeAsync(MerkleNode root) {
-            }
-            return result;
+            resultFuture = new FutureMerkleHash();
+            ResultTask resultTask = new ResultTask(root, resultFuture);
+            new TraverseTask(root, resultTask).send();
+            return resultFuture;
         }
     }
 
-    /**
-     * Create a thread that will attempt to hash the tree starting at the root.
-     *
-     * @param threadId
-     * 		Used to generate a randomized iteration order
-     */
-    private Runnable createHashingRunnable(
-            final int threadId, AtomicInteger activeThreadCount, final FutureMerkleHash result, final MerkleNode root) {
-
-        return () -> {
-            final MerkleIterator<MerkleNode> it = root.treeIterator()
-                    .setFilter(MerkleHashBuilder::filter)
-                    .setDescendantFilter(MerkleHashBuilder::descendantFilter);
-            if (threadId > 0) {
-                // One thread can iterate in-order, all others will use random order.
-                it.setOrder(POST_ORDERED_DEPTH_FIRST_RANDOM);
-            }
-
-            try {
-                hashSubtree(it, activeThreadCount);
-
-                // The last thread to terminate is responsible for setting the future
-                int remainingActiveThreads = activeThreadCount.getAndDecrement() - 1;
-                if (remainingActiveThreads == 0) {
-                    result.set(root.getHash());
-                }
-            } catch (final Throwable t) {
-                result.cancelWithException(t);
-            }
-        };
-    }
+    FutureMerkleHash resultFuture;
 
     /**
      * The root of a merkle tree.
      *
      * @param it
      * 		An iterator that walks through the tree.
-     * @param activeThreadCount
-     * 		If single threaded then this is null, otherwise contains the number of threads
-     * 		that are actively hashing. Once the active thread count dips below the maximum,
-     * 		this means that one thread has either finished or exploded.
      */
-    private void hashSubtree(final Iterator<MerkleNode> it, final AtomicInteger activeThreadCount) {
+    private void hashSubtree(final Iterator<MerkleNode> it) {
         while (it.hasNext()) {
             final MerkleNode node = it.next();
-            // Potential optimization: if this node is currently locked, do not wait here. Skip it and continue.
-            // This would require a lock object that support the "try lock" paradigm.
-            synchronized (node) {
-                if (activeThreadCount != null && activeThreadCount.get() != cpuThreadCount) {
-                    break;
-                }
 
-                if (node.getHash() != null) {
-                    continue;
-                }
+            if (node.getHash() != null) {
+                continue;
+            }
+
+            merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
+        }
+    }
+
+    class TraverseTask extends AbstractTask {
```

Comment:
Some docs would be helpful here. Maybe something simple like "This task is used to compute a hash of the given node and notify 'out' after that. If the node is a leaf or empty internal, the hash is computed using digestSync(), otherwise for every child node a new TraverseTask is created, ..."

Reply:
Done